base/abstractarray.jl

```diff
@@ -62,7 +62,11 @@ ndims{T<:AbstractArray}(::Type{T}) = ndims(super(T))
 length(t::AbstractArray) = prod(size(t))::Int
 endof(a::AbstractArray) = length(a)
 first(a::AbstractArray) = a[1]
-first(a) = (s = start(a); done(a, s) ? throw(ArgumentError("collection must be non-empty")) : next(a, s)[1])
+function first(arr)
```

Comment:
This is rather pedantic, I know, but I'd prefer the argument here to remain as `a` and not `arr`.  This definition *isn't* just for arrays — it's defined for all iterables.  Naming it `arr` strongly implies that it's just for arrays.  If you really want to steer away from the one-letter name, `itr` would be better.

Similarly, `s` is commonly used as the iterable's state throughout Julia.  `str` reads as string to me.